crates/autopilot/src/on_settlement_event_updater.rs

```diff
@@ -55,6 +56,10 @@ struct Inner {
     persistence: infra::Persistence,
     db: Postgres,
     notify: Notify,
+    /// Can be monitored to get notified about processed events.
+    updates_receiver: watch::Receiver<()>,
+    /// Can be used to notify about new processed events.
+    updates_sender: watch::Sender<()>,
```

Comment:
Having these different senders and receivers being subscribed from and triggered by very different parts of the code base makes things very hard to reason about imo (e.g. it's not clear to me what would happen if the settlement was included before the deadline, would we wait for another block in the runloop before starting a new auction?).

Can't we make everything synchronous in the runloop (and have the only condition that  on the block number must have changed compared to the last runloop)? Is it really necessary to have all these updates happen in a separate tasks that need trigger different synchronisation barriers? 

The runloop could imo simply be:
1. If there has been a previous auction with block x, wait for block x it to change (will most often be immediately the case)
2. [update_events](https://github.com/cowprotocol/services/blob/dadf16c0b5aadc7f72718dbfa3a9828d12b242cc/crates/shared/src/event_handling.rs#L254), this ensures all settlements are processed for all blocks until the current one.
3. [update](https://github.com/cowprotocol/services/blob/dadf16c0b5aadc7f72718dbfa3a9828d12b242cc/crates/autopilot/src/solvable_orders.rs#L169) the SolvableOrdersCache (this creates an auction with latest state applied)
4. Let solvers solve + winner settle
5. Repeat...

In terms of latency this should be the same as what is proposed here (we are doing the same amount of async work) but the code path will be much simpler to reason about.

Reply:
This has a problem when a solver wins an auction and then returns from the `/settle` call that it stopped trying the solution on-chain. In that case the start of the next auction might happen in the middle between 2 blocks. But that can also be solved with local variables (just thought about that 🤦 )
Overall I agree that it would be ideal to not jump through the hoops with the `OnEventUpdater` signal but I'm worried that this would be a significant refactor which would delay getting valuable performance measurements that could be worked on in parallel.

Indexing events becomes a lot less straight forward because we still need to be able to index historic events. One idea how to deal with I sketched out [here](https://github.com/cowprotocol/services/pull/2900#discussion_r1719448355) but this doesn't really work since it can leave the DB in an inconsistent state when there are historic events to be indexed, but the runloop already indexed a few more current events and the service getting restarted.
In that case we'd have a gap of unindexed events but we would not reindex this gap on the next restart because we already have more recent events indexed and we always start at the latest block we have any event for.

After thinking about this a bit more to me it seems indexing would need to be redesigned with these requirements:
1. on restart index all missing historic events (delays first run loop but shouldn't be too bad)
2. while waiting for the tx to land index all events we missed in the meantime
3. find tx on-chain and start next loop

Reply:
> In that case the start of the next auction might happen in the middle between 2 blocks.

Ok, so instead of the only condition being the block changed, we can also say something like "less than x seconds passed since the current block's timestamp". This way, if we would start an auction too late we rather skip and wait to start the next one on time.

 > Indexing events becomes a lot less straight forward because we still need to be able to index historic events.

Currently indexing events is [triggered](https://github.com/cowprotocol/services/blob/dadf16c0b5aadc7f72718dbfa3a9828d12b242cc/crates/shared/src/event_handling.rs#L506) from a maintenance task (similar to how solvable order updates is triggered from a background task). Why can we not in case of mode `SyncToBlockchain` call `update_events` from within the runloop? This will index all historic events and avoid ending up in inconsistent state. Most of the time `update_events` only processes a single block. In the case of a backlog there won't be any auctions but this might be ok?